crate-tmp/src/nvtxw.rs

```diff
@@ -0,0 +1,248 @@
+) -> Result<SessionHandle, ResultCode> {
+    let func = interface
+        .SessionBegin
+        .ok_or(NVTXW3_RESULT_INVALID_ARGUMENT)?;
+
+    let mut c_output: CString = CString::new("").expect("CString::new failed");
+    if let Some(os_output) = o_output {
+        let s_output = os_output.to_string_lossy();
+        let session_name = match s_output.strip_suffix(".nsys-rep") {
+            Some(s) => s,
+            None => &s_output,
+        };
+        c_output = CString::new(session_name).expect("CString::new failed");
+    }
+
+    let mut c_config: CString = CString::new("").expect("CString::new failed");
+    if let Some(os_merge) = o_merge {
+        let s_merge = os_merge.to_string_lossy();
+        let s_config = format!("ReportMerge={}", s_merge);
+
+        c_config = CString::new(s_config).expect("CString::new failed");
+    }
+
+    let session_attr = SessionAttributes {
+        struct_size: size_of::<SessionAttributes>(),
+        name: if c_output.is_empty() {
+            null_mut()
+        } else {
+            c_output.as_ptr()
+        },
+        configString: if c_config.is_empty() {
+            null_mut()
+        } else {
+            c_config.as_ptr()
+        },
+    };
+
+    let mut session: SessionHandle = Default::default();
+    check(unsafe { func(&mut session, &session_attr) })?;
+
+    if session.opaque.is_null() {
+        return Err(NVTXW3_RESULT_FAILED);
+    }
+
+    Ok(session)
+}
+
+pub fn stream_open_simple(
+    interface: InterfaceHandle,
+    session: SessionHandle,
+    stream_name: String,
+    domain_name: String,
+) -> Result<StreamHandle, ResultCode> {
+    let func = interface.StreamOpen.ok_or(NVTXW3_RESULT_INVALID_ARGUMENT)?;
+
+    let c_stream_name = CString::new(stream_name).expect("CString::new failed");
+    let c_domain_name = CString::new(domain_name).expect("CString::new failed");
+
+    let stream_attr = StreamAttributes {
+        struct_size: size_of::<StreamAttributes>(),
+        name: c_stream_name.as_ptr(),
+        nvtxDomainName: c_domain_name.as_ptr(),
+        eventScopePath: null(),
+        orderInterleaving: NVTXW3_STREAM_ORDER_INTERLEAVING_NONE,
+        orderingType: NVTXW3_STREAM_ORDERING_TYPE_UNKNOWN,
+        orderingSkid: NVTXW3_STREAM_ORDERING_SKID_NONE,
+        orderingSkidAmount: 0,
+    };
+
+    let mut stream: StreamHandle = Default::default();
+    check(unsafe { func(&mut stream, session, &stream_attr) })?;
+
+    if stream.opaque.is_null() {
+        return Err(NVTXW3_RESULT_FAILED);
+    }
+
+    Ok(stream)
+}
+
+pub fn schema_register(
+    interface: InterfaceHandle,
+    stream: StreamHandle,
+    schema_attr: &PayloadSchemaAttr,
+) -> Result<(), ResultCode> {
+    let func = interface
+        .SchemaRegister
+        .ok_or(NVTXW3_RESULT_INVALID_ARGUMENT)?;
+
+    check(unsafe { func(stream, schema_attr) })
+}
+
+pub fn event_write(
+    interface: InterfaceHandle,
+    stream: StreamHandle,
+    events: &[PayloadData],
```

Comment:
I'm sort of torn about this. On the one hand, I can see you're doing this to avoid copying data. On the other hand, it exposes the guts of `PayloadData` and the need to properly encode all the right C data structures to the user. This makes for fairly non-idiomatic Rust code on the end-user side.

I guess it's fine for now, but it seems like a prime candidate for making a better abstraction down the line.

Reply:
I agree that the resulting user code is not ideal. My understanding is that there is room for improvement by using Rust's macro system to populate some details of our structures, as long as your structure already has a C representation.